Start micro-frontend article (#230)
sogilis/Blog: site/content/posts/2021-01-08-micro-frontend.md

---
title: Micro frontend
author: Benoit et Willy
date: 2021-01-08
image: /img/2021-01-08-micro-frontend/microfrontend-thumbnail.jpeg
altimage: "'Crayons de couleur' by Supernico26 is licensed under CC BY 2.0"
categories:
    - FRONTEND
tags:
    - micro-frontend
    - micro
    - frontend
    - front
---

> L'architecture micro-service a rapidement été adoptée pour les services backend des applications web, typiquement en découpant ces services en différents blocs métiers. Généralement, ces services reposent sur une implémentation de type REST, qui permet un découplage complet entre la partie cliente (Web User Interface) d’un produit, et la partie backend qui gère les données.

Reviewer: ```suggestion
L'architecture micro-service a rapidement été adoptée pour les services backend des applications web, typiquement en découpant ces services en différents blocs métiers. Généralement, ces services reposent sur une implémentation de type REST. Ceci  permet un découplage complet entre la partie cliente (Web User Interface) d’un produit et la partie backend qui gère les données.
```

Cependant, avant d’utiliser un composant front de ce styleguide, il va devoir être pensé, puis créer, remanier, ou repenser encore. Difficile alors d’attendre qu’une autre équipe mette en place un composant, de comprendre comment il fonctionne, pour ensuite l’implémenter ailleurs où là base de code est peut être complètement différente.
Il est donc préférable de remplir encore la case de l’autonomie d’équipe, et de les laisser créer leurs propres composants chez eux au fur et à mesure qu’ils en ont besoin.
Il peut y avoir des duplications, mais une fois qu’elles sont devenues évidentes au niveau du modèle, c’est à ce moment-là qu’il est intéressant de récolter tout ça et de l’inscrire dans une bibliothèque partagée.
Ici, on parle bien uniquement de composants qui ont une **logique UI** (icônes, étiquettes, boutons, fonts, grilles, champs de recherche etc…). La bibliothèque ne doit pas contenir de logique métier (ou de domaine). Ces dernières appartiennent aux codes métiers du micro-frontend, pas à la bibliothèque.

> Émerge de cela un modèle de bibliothèque de composants avec un **gardien** ou **guilde** (une personne ou une équipe) qui va arbitrer et cultiver l’outil pour que la collaboration front entre les équipes se déroule bien. Il faut que les contributions à cette bibliothèque soient de bonne qualité, avec de la cohérence, et une bonne validité. Cela permet de consommer l’outil convenablement sans que les équipes se marchent dessus.

Reviewer 2: ```suggestion
Émerge de cela un modèle de bibliothèque de composants avec un **gardien** ou **guilde** (une personne ou une équipe) qui va arbitrer et cultiver l’outil pour que la collaboration front entre les équipes se déroule bien. Il faut que les contributions à cette bibliothèque soient de bonne qualité, avec de la cohérence, et une bonne validité. Cela permet de consommer l’outil convenablement sans que les équipes ne se marchent dessus.
```

## Charte graphique commune entre micro-frontends

> Le cas d'une bibliothèque de composants partagés est recommandable dans de grosses applications, avec plusieurs pages, et où il est possible d'identifier plusieurs composants pouvant être réutilisés de partout.

Reviewer 2: ```suggestion
Une bibliothèque de composants partagés est recommandée sur de grosses applications, avec plusieurs pages, et où il est possible d'identifier plusieurs composants pouvant être réutilisés de à plusieurs endroits.
```

Prenons l'exemple d'un site de commerce de produit en tout genre. La charte graphique a été réalisée par une équipe d'UX/UI et a été fournie à toutes les équipes de développement.

En terme de micro-frontends, nous pouvons identifier :
- une équipe A qui s'occupe de la page produit
- une équipe B qui s'occupe de la page d'accueil affichant tous les produits.
- une équipe C en charge du processus de panier/paiement

> De plus, sur la page produit, il est indiqué dans la charte d'afficher une liste de produit similaire.

Reviewer 2: Une charte graphique contient ce genre d'info ?

Reviewer 2: ```suggestion
De plus, sur la page produit, il est indiqué dans la charte d'afficher une liste de produits similaires.
```

> Afficher un produit revient alors à afficher un bloc contenant son image, son intitulé, son prix et un lien dirigeant vers sa propre page.

Reviewer 2: Je ne vois pas le lien avec la liste des produits similaires (rapport au mot "alors" employé ici).

> La façon d'afficher ces produits va être conduit par l'équipe A qui en a besoin pour sa page d'accueil, et par l'équipe B qui en a besoin aussi pour les produits similaires.

Reviewer 2: ```suggestion
La façon d'afficher ces produits va être conduite par l'équipe A qui en a besoin pour sa page d'accueil, et par l'équipe B qui en a besoin aussi pour les produits similaires.
```

> Chacune des équipes va donc créer son propre composant produit, avec ses propres règles. Ensuite, une équipe plus annexe, va avoir la responsabilité de fusionner les règles du composant définient par l'équipe A et B en respectant la charte initiale. Cette fusion va donner aux équipes de développement un composant produit de référence, ranger dans une bibliothèque qui leurs est accessible.

Reviewer 2: ```suggestion
Chacune des équipes va donc créer son propre composant produit, avec ses propres règles. Ensuite, une équipe annexe, a la responsabilité de fusionner les règles du composant définies par les équipes A et B en respectant la charte initiale. Cette fusion va donner aux équipes de développement un composant produit de référence, disponible dans une bibliothèque qui leurs est accessible.
```

Cette équipe annexe n'est même pas dans l'obligation de fournir un composant de référence dans une bibliothèque, mais peut simplement emettre aux équipes de développement les règles CSS (padding, marging, fonts, taille etc...) qui doivent être respectées.

## L'anarchie des micro-frontends

> De plus en plus, l'architecture en micro-frontend gagne en popularité. Mais elle est parfois mal utilisée, voir utilisée avec abus. Un biais commun, est de mélanger les technologies et les outils, laissant lieu à une complexité indéniable. Un exemple peut être l'utilisation de plusieurs frameworks structurants, comme React, Vue, Angular, ou encore diversifier à outrance la gestion du style (une application avec SASS, l'autre en BEM, ou encore avec du CSS-in-JS). Tout cela déteriore la cohérence technique de l'application globale.

Reviewer 2: ```suggestion
L'architecture en micro-frontend gagne de plus en plus en popularité. Mais elle est parfois mal utilisée, voir utilisée avec abus. Un biais commun, est le mélange de technologies et d'outils, laissant place à une complexité indéniable comme par exemple l'utilisation de plusieurs frameworks structurants (`React`, `Vue`, `Angular`) ou diversifier à outrance la gestion du style (une application avec SASS, l'autre en BEM, ou encore avec du CSS-in-JS). Tout cela déteriore la cohérence technique de l'application globale.
```

> Là où les équipes sont complètement autonomne dans un système micro-frontend, il faut quand même essayer de standardiser les approches et les technologies utilisées.

Reviewer 2: ```suggestion
Même si les équipes sont complètement autonomes dans un système micro-frontend, il faut tout de même essayer de standardiser les approches et les technologies utilisées.
```

---
title: Micro frontend
author: Benoit et Willy
date: 2021-01-08
image: /img/2021-01-08-micro-frontend/microfrontend-thumbnail.jpeg
altimage: "'Crayons de couleur' by Supernico26 is licensed under CC BY 2.0"
categories:
    - FRONTEND
tags:
    - micro-frontend
    - micro
    - frontend
    - front
---

> 

Reviewer 3: Hello, ci dessous un brouillon d'une intro à ajouter / merger avec autres choses ;) 

Dans le domaine du génie logiciel, les microservices désignent un type d'architecture dans lequel un ensemble de services indépendants et autonomes forment un tout cohérent.
Contrairement à une approche monolithique dans laquelle tous les éléments du logiciel sont étroitement liés, l'approche micro service consiste à fragmenter la logique en de multiples composants (écrits/basés sur différentes technologies) qui communiquent via des API définies.

Du point de vue des développeurs, l'idée principale repose dans la possibilité de gérer chaque service comme un projet à part entière, ce qui en facilite entre autres, la maintenance, l'évolution et la gestion projet.
En effet, une telle approche permet une répartition des différents travaux entre les équipes tout en leur accordant autonomie et liberté concernant leur organisation et leurs choix des technologies.

Dans le domaine du web l'approche microservice est souvent utilisée dans le cadre de l'implémentation de backends. Ceux-ci fonctionnent communément sous la forme d'un ensemble d'API REST qui discutent entre elles et / ou répondent aux requêtes des interfaces utilisateurs.
De manière similaire, la conception d'interfaces graphiques peut suivre une logique analogue. L'idée principale réside sur le principe de composition. En effet il s'agit d'une architecture dans laquelle des **sous-produits fronts indépendants** sont **assemblés pour former un tout plus important, une application**.

L'approche microservice appliquée au frontend et au backend permet de confier la réalisation d'une fonctionnalité à une équipe autonome et polyvalente chargée de sa conception dans son ensemble, de la gestion des données à leur présentation.
Dans le cadre de l'élaboration d'un logiciel, l'autonomie des équipes peut être une source de conflits et difficultés. C'est la raison pour laquelle il est nécessaire de définir un cadre et des règles de construction claires afin de coordonner et synchroniser les travaux afin d'aboutir à un produit cohérent.

Ce document se focalise sur l'application d'une approche microservice du côté du frontend. Avec quelles méthodes / outils les équipes peuvent-elles coordonner leur travaux et garantir la livraison d'un produit fonctionnel ?
En effet la cohérence des styles, l'utilisation de bibliothèques partagées et d'une charte graphique sont autant de modes de fonctionnement en faveur d'une appréhension et application correcte du paradigme.

L'architecture micro-service a rapidement été adoptée pour les services backend des applications web, typiquement en découpant ces services en différents blocs métiers. Généralement, ces services reposent sur une implémentation de type REST. Ceci  permet un découplage complet entre la partie cliente (Web User Interface) d’un produit et la partie backend qui gère les données.
Cependant, et grâce au découplage apporté par l'intertace REST, ce découpage en micro-services backend n'influence en rien l'architecture de la partie cliente (ou front).
Cette problématique est aussi présente sur des applications frontend monolothique. C'est encore l'architecture que l'on retrouve malheuresement le plus souvent dans les logiciels Web.

Le micro-frontend propose alors une approche différente. C’est un style d’architecture dans lequel des **sous-produits fronts indépendants** sont **composées ensemble pour former un tout plus grand**.

> Par exemple, avec l’approche DDD (Domain Driven Design), on peut voir les sous-produits comme des "sous-domaines" (bounded context) d’un domaine.

Reviewer 3: Il pourrait être interessant de clarifier la notion de Domain Driven Design en mettant un lien vers la page [wikipedia](https://fr.wikipedia.org/wiki/Conception_pilot%C3%A9e_par_le_domaine) ou de la [dddCommunity](https://www.dddcommunity.org/learning-ddd/what_is_ddd/) ;)

Quel est le coût d’une telle architecture ? Il peut y avoir des fragmentations dans la manière dont les équipes travaillent et qui peut être causée par l’autonomie qui leurs sont accordées.
Plus les équipes augmentent, plus il devient difficile de s’assurer que les connaissances sont transférées et que le code ne devienne pas trop hétérogène entre les équipes.

## Comment gérer la cohérence du style ?

Si dans un micro-frontend A, la feuille de style dit `a { color: red; }` et que celle du micro-frontend B dit que `a { color: blue; }`, une des deux équipes sera forcément déçue.
Dans ce mode d’architecture, et par le fait que le CSS est un langage héréditaire et en cascade, si tous les sélecteurs écrits par les équipes sont rassemblées en une page, un problème se pose. Comment rendre le CSS plus maniable ?

Le plus important est de trouver un moyen de s’assurer que les développeurs codent leurs styles indépendamment les uns des autres, et que la composition de ces styles en une seule application est prévisible.
Depuis quelques années, plusieurs approches ont été inventées pour faire cela :

- avoir une convention de dénomination stricte :

> par exemple avec la [syntaxe BEM](http://getbem.com/) qui s’assure que les sélecteurs sont appliqués seulement aux endroits prévus (dans un composant par exemple)

Reviewer 3: Double répétition de "par exemple". 
Je pense que : "La [syntaxe BEM](http://getbem.com/) qui s’assure que les sélecteurs sont appliqués seulement aux endroits prévus (dans le cadre d'un composant par exemple) et avec une syntaxe précise" devrait suffire !

> - l’utilisation d’un préprocesseur, comme [SASS](https://sass-lang.com/) ou [LESS](http://lesscss.org/)

Reviewer 3: Il pourrait être intéressant de préciser (en une petite phrase, comme pour BEM) en quoi l'utilisation d'un préprocesseur peut être une solution. 
En effet ce type de préprocesseur permet de structurer / hiérarchiser le css ce qui a pour effet de limiter les effets de bords.

> - appliquer les styles programmatiquement avec des modules CSS ou avec du CSS-in-JS, pour garantir que les styles sont appliqués juste au bon endroit

Reviewer 3: ```suggestion
- appliquer les styles programmatiquement avec les [Css modules](https://github.com/css-modules/css-modules) ou avec du [CSS-in-JS ](https://github.com/css-modules/css-modules), pour garantir que les styles sont appliqués aux bons endroits
```

> - l’utilisation du [shadow DOM](https://developer.mozilla.org/fr/docs/Web/Web_Components/Using_shadow_DOM) avec des [Web Component](https://developer.mozilla.org/fr/docs/Web/Web_Components) par exemple

Reviewer 3: Pareil que ci dessus, une petite phrase d'explication serait vraiment top :)

> ## Peut-on utiliser une bibliothèque de composants partagés

Reviewer 3: ```suggestion
## Peut-on utiliser une bibliothèque de composants partagés ?
```

- une équipe B qui s'occupe de la page d'accueil affichant tous les produits.
- une équipe C en charge du processus de panier/paiement

De plus, sur la page produit, il est indiqué dans la maquette d'afficher une liste de produits similaires.

Afficher un produit revient à afficher un bloc contenant son image, son intitulé, son prix et un lien dirigeant vers sa propre page.

La façon d'afficher ces produits va être conduit par l'équipe A qui en a besoin pour sa page d'accueil, et par l'équipe B qui en a besoin aussi pour les produits similaires.
Chacune des équipes va donc créer son propre composant produit, avec ses propres règles. Ensuite, une équipe annexe, a la responsabilité de fusionner les règles du composant définies par les équipes A et B en respectant la charte initiale. Cette fusion va donner aux équipes de développement un composant produit de référence, disponible dans une bibliothèque qui leurs est accessible.
Cette équipe annexe n'est même pas dans l'obligation de fournir un composant de référence dans une bibliothèque, mais peut simplement emettre aux équipes de développement les règles CSS (padding, marging, fonts, taille etc...) qui doivent être respectées.

> ## L'anarchie des micro-frontends

Reviewer 3: Sans rentrer dans le débat de l'anarchisme, je pense que le terme à utiliser ici est "Chaos" ("Le chaos des micro-frontends") ;)

L'architecture en micro-frontend gagne de plus en plus en popularité. Mais elle est parfois mal utilisée, voir utilisée avec abus. Un biais commun, est le mélange de technologies et d'outils, laissant place à une complexité indéniable comme par exemple l'utilisation de plusieurs frameworks structurants (`React`, `Vue`, `Angular`) ou diversifier à outrance la gestion du style (une application avec SASS, l'autre en BEM, ou encore avec du CSS-in-JS). Tout cela déteriore la cohérence technique de l'application globale.
Même si les équipes sont complètement autonomes dans un système micro-frontend, il faut tout de même essayer de standardiser les approches et les technologies utilisées.

> 

Reviewer 3: Il manque une petite conclusion en lien avec l'introduction / problématique afin de terminer proprement le papier :)

Reviewer 3: Une architecture microservice permet d'une part l'utilisation et la cohabitation de technologies et d'outils hétérogènes, d'autre part, un fonctionnement sous forme de multiples équipes polyvalentes et autonomes.
Concevoir une application web sous la forme de microservices développés par plusieurs équipes nécessite une harmonisation et une spécification des pratiques.

Du point de vue front, le développement doit s'effectuer dans le respect de règles claires et tout particulièrement en ce qui concerne la cohérence des règles css. En effet des conventions de nommage ou l'utilisation d'outils dédiés permettent d'écarter en grande partie les conflits liés aux règles.
L'utilisation d'une bibliothèque de composant commune aux équipes apparait en faveur de la conservation d'une unité graphique. Celle-ci peut par ailleurs être améliorée au fur à mesure des développements afin d'être adaptée aux besoins.
Enfin, confier la réalisation d'une charte graphique ainsi que le maquettage à une équipe dédiée (UX / UI) permet d'obtenir des directives claires qui permettent à chaque équipe de développeurs de produire leurs services de manière cohérente.
Le succès de la conception d'un produit avec une architecture microservices ne repose pas exclusivement sur la qualité des outils et technologies utilisés. Mais dépend principalement des modes de fonctionnement et d'organisation permettant la bonne coopération des équipes.

---
title: Micro frontend
author: Benoit et Willy
date: 2021-01-08
image:
categories:
    - FRONTEND
tags:
    - micro-frontend
    - micro
    - frontend
    - front
---

> Le mode de développement en micro-service a rapidement été adopté pour les services backend des applications web, typiquement via un découpage des services en blocs métiers. La préconisation pour la réalisation de ces services est une implémentation de type REST, qui permet un découplage complet entre la partie cliente (Web User Interface) d’un produit, et la partie backend qui gère les données.

Reviewer 4: ```suggestion
L'architecture micro-service a rapidement été adoptée pour les services backend des applications web, typiquement en découpant ces services en différents blocs métiers. Généralement, ces services reposent sur une implémentation de type REST, qui permet un découplage complet entre la partie cliente (Web User Interface) d’un produit, et la partie backend qui gère les données.
```

> Malgré ce découpage en micro-services backend, la possibilité d’une application cliente (ou front) monolithique était toujours possible.

Reviewer 4: ```suggestion
Cependant, et grâce au découplage apporté par l'intertace REST, ce découpage en micro-services backend n'influence en rien l'architecture de la partie cliente (ou front). Aussi, une application frontend monolithique s'interface très bien avec de multiples services backend, et malheureusement, c'est encore l'architecture que l'on retrouve le plus souvent dans les logiciels Web.
```

> Le micro-frontend vient alors en aide. C’est un style d’architecture dans lequel des **sous-produits fronts indépendants** sont **composées ensemble pour former un tout plus grand**.

Reviewer 4: ```suggestion
Le micro-frontend vient alors en aide. C’est un style d’architecture dans lequel des **sous-produits fronts indépendants** sont **composées ensemble pour former un tout cohérent**.
```

> Si l’approche DDD (Domain Driven Design), voyez les sous-produits comme les sous-domaines d’un domaine.

Reviewer 4: ```suggestion
Par exemple, avec l’approche DDD (Domain Driven Design), on peut voir les sous-produits comme des sous-domaines d’un domaine.
```

Reviewer 4: Je ne suis pas super convaincu de l'intérêt , voir de la véracité de cette phrase. On peut éventuellement la supprimer simplement.

> Pour cela, chaque sous-produit d’un produit applicatif est sous la responsabilité **d’une seule équipe fullstack autonome**, et n’a ainsi pas besoin de se coordonner avec quelqu’un d’autre pour le mettre en production.

Reviewer 4: ```suggestion
Dans une architecture micro-frontend, chaque sous-produit est indépendant, et peut potentiellement être sous la responsabilité **d’une seule équipe autonome**. L'équipe en charge de déveloper ce sous-produit, n’a ainsi pas besoin de se coordonner avec les autres équipes pour gérer son cycle de vie (déploiement, mise en production, mise à jours, etc.). Toute problématique d'intégration avec les autres sous-produits et les équipes qui en ont la charge est alors éliminée.
```

> Ainsi, celle-ci a donc la responsabilité **du backend et du frontend**, on y trouve donc des développeurs fullstack et/ou des développeurs backend et/ou des développeurs frontend.

Reviewer 4: ```suggestion
Avec ce modéle, une équipe a la responsabilité **du backend et du frontend**, on y trouve donc des développeurs fullstack et/ou des développeurs backend et/ou des développeurs frontend.
```

Reviewer 3: "modèle" ;)

Dans cette architecture, le plus important va être de livrer de bout-en-bout en étant totalement indépendant.

![micro-frontend](/img/2021-01-08-micro-frontend/micro-frontend.png)

> Les premiers avantages à relevés sont « l'indépendance opérationnelle » ou « l’indépendance au runtime » :

Reviewer 4: ```suggestion
Les premiers avantages à relever sont « l'indépendance opérationnelle » ou « l’indépendance au runtime » :
```

> - généralement l’équipe à son propre dépôt de code source et peut déployer ses services indépendamment les uns des autres, en choisissant les technologies qu’elle souhaite mettre en place (exemple : Go pour le backend et ReactJS pour le frontend pour le sous-produit A).

Reviewer 4: ```suggestion
- généralement l’équipe a son propre dépôt de code source et peut déployer ses services indépendamment les uns des autres, en choisissant les technologies qu’elle souhaite mettre en place (exemple : Go pour le backend et ReactJS pour le frontend pour le sous-produit A).
```

> - l’équipe à sa propre suite de tests automatisés et une pipeline de livraison vers la production.

Reviewer 4: ```suggestion
- l’équipe a sa propre suite de tests, de préférence automatisés, et son propre pipeline de livraison vers la production.
```

> De cette manière, une seule équipe peut prendre par exemple la charge d’ajout d’une fonctionnalité au produit, et l’amener par ses propres moyens vers une livraison finale, sans impacter les autres équipes. Le tout de manière fullstack.

Reviewer 4: ```suggestion
De cette manière, les équipes son indépendantes. Chacune peut ajouter des fonctionnalités au produit, et les amener par ses propres moyens, vers une livraison en production et, le tout, sans impacter les autres équipes.
```
J'enléverrai « de manière fullstack ».

> Le produit général se retrouve avec des bases de codes plus petites, plus cohérentes et maintenables. Mais aussi avec des équipes découplées et autonomes permettant de gagner en évolutivité. Tout cela favorise une mise à jour, une mise à niveau ou une réécriture des parties du frontend, bien plus progressive qu’auparavant.

Reviewer 4: ```suggestion
Le produit général se retrouve avec des bases de code plus petites, plus cohérentes et maintenables. Mais aussi avec des équipes découplées et autonomes permettant de gagner en évolutivité. Tout cela favorise une mise à jour, une mise à niveau ou une réécriture des parties du frontend, bien plus progressive qu’avec une architecture monolitique.
```

> Quel est le coût d’une telle architecture ? Il peut y avoir des fragmentations dans la manière dont les équipes travaillent et qui peut être causée par l’autonomie qui leurs sont accordées.

Reviewer 4: On peut expliquer « fragmentations » ? 

Je ne suis pas sûr de comprendre.

> Plus les équipes augmentent, plus il devient difficile de s’assurer que les connaissances sont transférées et que les équipes ne deviennent pas cloisonnées.

Reviewer 4: ```suggestion
Plus le nombre d'équipes augmente, plus il devient difficile de s’assurer que les connaissances sont transférées et que les équipes ne deviennent pas cloisonnées.
```

> ## Comment gérer la cohérence du style ?

Reviewer 4: ```suggestion
## Comment gérer la cohérence du style ?
Dans une architecture micro-frontent, avec laquelle des équipes indépendantes développent des fonctionnalités, on peut voir apparaître une problèmatique de cohérence du style.

Voyons un exemple.
```

> Si dans un micro-frontend A, la feuille de style dit `a { color: red; }` et que celle du micro-frontend B dit que `a { color: blue; }`, une des deux équipes sera forcément déçue.

Reviewer 4: ```suggestion
Si dans un micro-frontend A, la feuille de style dit `a { color: red; }` et que celle du micro-frontend B dit que `a { color: blue; }`, les équipes sont face a un problème de cohérence de style.
```

> Dans ce mode d’architecture, et par le fait que le CSS est un langage héréditaire et en cascade, si tous les sélecteurs écrits par les équipes sont rassemblées en une page, un problème se pose. Comment rendre le CSS plus maniable ?

Reviewer 4: Est-ce qu'on expliquerait pas, en une phrase, le problème qui se pose ? 

Il est recommandé de ne rien laisser d'implicite à la rédaction d'un article (des gens ne maitrisant pas le CSS pourrait avoir envie de comprendre).

> Le plus important est de trouver un moyen de s’assurer que les développeurs codent leurs styles indépendamment les uns des autres, et que la composition de ces styles en une seule application est prévisible.

Reviewer 4: ```suggestion
Le plus important est de trouver un moyen de s’assurer que les développeurs codent leurs styles indépendamment les uns des autres, et que la composition de ces styles en une seule application est prévisible et cohérente.
```

Depuis quelques années, plusieurs approches ont été inventées pour faire cela :

- avoir une convention de dénomination stricte :
par exemple avec la [syntaxe BEM](http://getbem.com/) qui s’assure que les sélecteurs sont appliqués seulement aux endroits prévus (dans un composant par exemple)

- l’utilisation d’un préprocesseur, comme [SASS](https://sass-lang.com/) ou [LESS](http://lesscss.org/)

- appliquer les styles programmatiquement avec des modules CSS ou avec du CSS-in-JS, pour garantir que les styles sont appliqués juste au bon endroit

- l’utilisation du [shadow DOM](https://developer.mozilla.org/fr/docs/Web/Web_Components/Using_shadow_DOM) avec des [Web Component](https://developer.mozilla.org/fr/docs/Web/Web_Components) par exemple

## Peut-on utiliser une bibliothèque de composants partagés

> Visuellement, il est clair qu’il faut mettre de l’importance sur la cohérence visuelle entre les micro-frontends. Une bibliothèque de composants UI partagés et réutilisables est en général une bonne préconisation. Cette bibliothèque de composants sert de guide de style (styleguide) collaboratif entre les acteurs du produit à développer.

Reviewer 4: ```suggestion
Il est clair qu’il faut mettre en avant la cohérence visuelle entre les micro-frontends. Une bibliothèque de composants UI partagés et réutilisables est une réelle préconisation. Cette bibliothèque de composants sert de guide de style (styleguide) collaboratif entre les acteurs du produit à développer.
```

> Cependant, avant d’utiliser un composant front de ce styleguide, il va devoir être pensé, puis créer, remanier, ou repenser encore. Difficile alors d’attendre qu’une autre équipe mette en place un composant, de comprendre comment il fonctionne, pour ensuite l’implémenter ailleurs où là base de code est peut être complètement différente.

Reviewer 4: ```suggestion
Cependant, avant d’utiliser un composant front de ce styleguide, il va devoir être pensé, créé, remanié, et repensé encore. Difficile alors d’attendre qu’une autre équipe mette en place ce composant. En effet, si une équipe A apporte des remaniements à ce composant, il va être difficile pour une équipe B de comprendre comment il fonctionne, pour ensuite l’implémenter, ailleurs, où là base de code est peut être complètement différente.
```

Il est donc préférable de remplir encore la case de l’autonomie d’équipe, et de les laisser créer leurs propres composants chez eux au fur et à mesure qu’ils en ont besoin.
Il peut y avoir des duplications, mais une fois qu’elles sont devenues évidentes au niveau du modèle, c’est à ce moment-là qu’il est intéressant de récolter tout ça et de l’inscrire dans une bibliothèque partagée.
Ici, on parle bien uniquement de composants qui ont une **logique UI** (icônes, étiquettes, boutons, fonts, grilles, champs de recherche etc…). La bibliothèque ne doit pas contenir de logique métier (ou de domaine). Ces dernières appartiennent aux codes métiers du micro-frontend, pas à la bibliothèque.

> Émerge de cela un modèle de bibliothèque de composants avec un **gardien** ou **guilde** (une personne ou une équipe) qui va arbitrer et cultiver l’outil pour que la collaboration front entre les équipes se déroule bien. Il faut que les contributions à cette bibliothèque soient de bonne qualité, avec de la cohérence, et une bonne validité. Cela permet de consommer l’outil convenablement sans que les équipes se marchent dessus.

Reviewer 4: ```suggestion
Émerge de cela un modèle de bibliothèque de composants avec un **gardien** ou **guilde** (une personne ou une équipe) qui va arbitrer et cultiver l’outil pour que la collaboration front entre les équipes se déroule bien. Il faut que les contributions à cette bibliothèque soient de bonne qualité, avec de la cohérence, et une bonne validité. Cela permet de d'utiliser l’outil convenablement sans que les équipes se marchent dessus.
```

Reviewer 2: Que veut dire "une bonne validité" ?

## Charte graphique commune entre micro-frontends

> Le cas d'une bibliothèque de composants partagées est pensable dans de grosses applications, avec plusieurs pages et où il est possible de visualiser plusieurs composants qui peuvent être réutilisés de partout.

Reviewer 4: ```suggestion
Le cas d'une bibliothèque de composants partagés est recommandable dans de grosses applications, avec plusieurs pages, et où il est possible d'identifier plusieurs composants pouvant être réutilisés de partout.
```

> Prenons l'exemple d'un site de commerce de produit en tout genre. La charte graphique a été réalisé par une équipe d'UX/UI et a été fourni à tous les équipes de développement.

Reviewer 4: ```suggestion
Prenons l'exemple d'un site de commerce de produit en tout genre. La charte graphique a été réalisée par une équipe d'UX/UI et a été fournie à toutes les équipes de développement.
```

En terme de micro-frontends, nous pouvons identifier :
- une équipe en charge du processus de panier/paiement
- une autre qui s'occupe de la page produit
- une autre de la page d'accueil qui affiche tous les produits.

De plus, sur la page produit, il est indiqué dans la charte d'afficher une liste de produit similaire.

> Afficher un produit revient alors à afficher son image, son intitulé, son prix et un lien dirigeant vers sa propre page. La façon d'afficher un produit va être conduit par l'équipe de la page d'accueil qui en a besoin et de celle de la page produit qui en a besoin aussi pour les produits similaires.

Reviewer 4: ```suggestion
Afficher un produit revient alors à afficher son image, son intitulé, son prix et un lien dirigeant vers sa propre page. La façon d'afficher un produit va être conduit par l'équipe de la page d'accueil, qui en a besoin, et de celle de la page produit, qui en a besoin aussi pour les produits similaires.
```

> Chacun va donc créer son composant avec les règles qui lui sont propres. Ensuite, plusieurs personnes ont la responsabilité de déployer ce composant dans une bibliothèque, en fusionnant les règles définies par les deux autres équipes, et en respectant la charte initiale. Il serira donc de composant de référence indiquant à toutes les équipes ses marges, padding, tailles, fonts etc...

Reviewer 4: ```suggestion
Chacun va donc créer son composant avec les règles qui lui sont propres. Ensuite, plusieurs personnes ont la responsabilité de déployer ce composant dans une bibliothèque, en fusionnant les règles définies par les deux autres équipes, et en respectant la charte initiale. Il servira donc de composant de référence indiquant à toutes les équipes ses marges, padding, tailles, fonts etc...
```
On peut retravailler un peu le texte je pense. L'idée est bonne, mais pas facile à expliquer

## L'anarchie des micro-frontends

> De plus en plus, les systèmes en micro-frontends gagnent en popularité. Mais ils sont parfois mal utilisée voir complètement abusée. Un biais serait de s'en service d'excuse pour mélanger les technologies et les outils, laissant lieu à une complexité indéniable. L'utilisation de plusieurs framework (ReactJS, VueJS, Angular), ou sur la manière dont le style est mis en place (une application avec SASS, l'autre en BEM, ou encore avec du CSS-in-JS) déteriore la cohérence technique de l'application globale.

Reviewer 4: ```suggestion
De plus en plus, l'architecture en micro-frontend gagne en popularité. Mais elle est parfois mal utilisée, voir utilisée avec abus. Un biais commun, est mélanger les technologies et les outils, laissant lieu à une complexité indéniable. Un exemple peut-être l'utilisation de plusieurs frameworks structurants, comme ReactJS, VueJS, Angular, ou encore diversifier à outrance la gestion du style (une application avec SASS, l'autre en BEM, ou encore avec du CSS-in-JS). Tout celà déteriore la cohérence technique de l'application globale.
```
